qr.c

```diff
@@ -0,0 +1,195 @@
+
+#include <stdlib.h>
+
+#include "main.h"
+#include "qr.h"
+#include "misc.h"
+
+#include "openvpn-gui-res.h"
+
+#define QR_SCALE           3
+
+/* used to close dialog from outside */
+#define WM_CLOSE_QR_DIALOG (WM_USER + 1)
+static HWND g_hwndQR = NULL;
+
+/* Generate QR bitmap from wide string */
+HBITMAP
+CreateQRBitmapFromWchar(const wchar_t *wtext, int scale)
+{
+    char *utf8_text = WCharToUTF8(wtext);
+    if (!utf8_text)
+        return NULL;
+
+    uint8_t qrcode[qrcodegen_BUFFER_LEN_MAX];
+    uint8_t temp[qrcodegen_BUFFER_LEN_MAX];
+
+    bool ok = qrcodegen_encodeText(utf8_text,
+                                   temp,
+                                   qrcode,
+                                   qrcodegen_Ecc_LOW,
+                                   qrcodegen_VERSION_MIN,
+                                   qrcodegen_VERSION_MAX,
+                                   qrcodegen_Mask_AUTO,
+                                   true);
+    free(utf8_text);
+    if (!ok)
+        return NULL;
+
+    int size = qrcodegen_getSize(qrcode);
+    int imgSize = size * scale;
+
+    HDC hdcScreen = GetDC(NULL);
+    HDC hdcMem = CreateCompatibleDC(hdcScreen);
+    HBITMAP hBitmap = CreateCompatibleBitmap(hdcScreen, imgSize, imgSize);
+    SelectObject(hdcMem, hBitmap);
+
+    RECT rc = { 0, 0, imgSize, imgSize };
+    FillRect(hdcMem, &rc, (HBRUSH)GetStockObject(WHITE_BRUSH));
+
+    HBRUSH blackBrush = (HBRUSH)GetStockObject(BLACK_BRUSH);
+    for (int y = 0; y < size; ++y)
+    {
+        for (int x = 0; x < size; ++x)
+        {
+            if (qrcodegen_getModule(qrcode, x, y))
+            {
+                RECT pixelRect = { x * scale, y * scale, (x + 1) * scale, (y + 1) * scale };
+                FillRect(hdcMem, &pixelRect, blackBrush);
+            }
+        }
+    }
+
+    DeleteDC(hdcMem);
+    ReleaseDC(NULL, hdcScreen);
+    return hBitmap;
+}
+
+INT_PTR CALLBACK
+QrDialogProc(HWND hwnd, UINT msg, UNUSED WPARAM wp, LPARAM lp)
+{
+    static HBITMAP hQRBitmap = NULL;
+
+    switch (msg)
+    {
+        case WM_INITDIALOG:
+        {
+            const wchar_t *qrText = (const wchar_t *)lp;
+            const int padding = 15;
+            const int qrMaxSize = 300; /* max QR bitmap size in pixels */
+
+            /* Generate QR initially at largest acceptable size */
+            int scale = 6;
+            HBITMAP hQRBitmap = CreateQRBitmapFromWchar(qrText, scale);
+
+            if (!hQRBitmap)
+            {
+                MsgToEventLog(EVENTLOG_ERROR_TYPE, L"Failed to generate QR");
+                EndDialog(hwnd, 0);
+                return TRUE;
+            }
+
+            BITMAP bm;
+            GetObject(hQRBitmap, sizeof(BITMAP), &bm);
+
+            /* Adjust scale down if bitmap too big */
+            while ((bm.bmWidth > qrMaxSize) && scale > 1)
+            {
+                DeleteObject(hQRBitmap);
+                scale--;
+                hQRBitmap = CreateQRBitmapFromWchar(qrText, scale);
```

Comment:
Better to check return value of CreateQRBitmapfromWchar (even if unlikely to fail at this point)